File: ios/MoniVitals/Model/Store.swift
//
//  Store.swift
//  MoniVitals
//
//  Dependency-free, local-first persistence (no cloud sync). Mirrors the responsibilities
//  of the web Dexie database (web/src/model/db.ts + repository.ts) but stores data as
//  JSON files under Application Support:
//
//      <AppSupport>/MoniVitals/
//        subjects.json          [Subject]
//        sessions.json          [Session]
//        calibrations.json      [Calibration]
//        annotations.json       [Annotation]
//        chunks/<sessionId>.json     [StreamChunk]   (high-rate sample blocks)
//        metrics/<sessionId>.json    [MetricsRecord]
//
//  Per-session sample data lives in its own file so listing/deleting sessions never
//  loads every waveform. All access is serialized on a private queue.
//

import Foundation

/// Concatenated, contiguous streams for a session (replay / export).
struct SessionStreams {
    struct Ecg { var t: [Double]; var ecg: [Int32] }
    struct Bioz { var t: [Double]; var z0: [Int32]; var dz: [Int32] }
    struct Ppg { var t: [Double]; var green: [UInt32]; var red: [UInt32]; var ir: [UInt32] }
    var ecg = Ecg(t: [], ecg: [])
    var bioz = Bioz(t: [], z0: [], dz: [])
    var ppg = Ppg(t: [], green: [], red: [], ir: [])
}

/// The local data store. A single shared instance is used by the app.
final class Store {
    static let shared = Store()

    private let queue = DispatchQueue(label: "com.monivitals.store")
    private let fm = FileManager.default
    private let root: URL
    private let chunksDir: URL
    private let metricsDir: URL
    private let encoder: JSONEncoder
    private let decoder = JSONDecoder()

    private init() {
        let base = (try? fm.url(for: .applicationSupportDirectory, in: .userDomainMask,
                                appropriateFor: nil, create: true))
            ?? fm.temporaryDirectory
        root = base.appendingPathComponent("MoniVitals", isDirectory: true)
        chunksDir = root.appendingPathComponent("chunks", isDirectory: true)
        metricsDir = root.appendingPathComponent("metrics", isDirectory: true)
        encoder = JSONEncoder()
        try? fm.createDirectory(at: root, withIntermediateDirectories: true)
        try? fm.createDirectory(at: chunksDir, withIntermediateDirectories: true)
        try? fm.createDirectory(at: metricsDir, withIntermediateDirectories: true)
    }

    // MARK: - file helpers

    private func url(_ name: String) -> URL { root.appendingPathComponent(name) }

    private func load<T: Decodable>(_ url: URL, default def: T) -> T {
        guard let data = try? Data(contentsOf: url) else { return def }
        return (try? decoder.decode(T.self, from: data)) ?? def
    }

    private func save<T: Encodable>(_ value: T, to url: URL) {
        if let data = try? encoder.encode(value) {
            try? data.write(to: url, options: .atomic)
        }
    }

    private func nextId(_ items: [Int?]) -> Int {
        (items.compactMap { $0 }.max() ?? 0) + 1
    }

    // MARK: - Subjects

    func upsertSubject(_ s: Subject) {
        queue.sync {
            var all: [Subject] = load(url("subjects.json"), default: [])
            all.removeAll { $0.code == s.code }
            all.append(s)
            save(all, to: url("subjects.json"))
        }
    }

    /// Subjects, newest first (by createdAt) — matches the web `listSubjects`.
    func listSubjects() -> [Subject] {
        queue.sync {
            let all: [Subject] = load(url("subjects.json"), default: [])
            return all.sorted { $0.createdAt > $1.createdAt }
        }
    }

    func getSubject(_ code: String) -> Subject? {
        queue.sync {
            let all: [Subject] = load(url("subjects.json"), default: [])
            return all.first { $0.code == code }
        }
    }

    func deleteSubject(_ code: String) {
        // Delete the subject's sessions + calibrations, then the subject.
        let sessions = listSessions().filter { $0.subjectCode == code }
        for s in sessions { deleteSession(s.id) }
        queue.sync {
            var cals: [Calibration] = load(url("calibrations.json"), default: [])
            cals.removeAll { $0.subjectCode == code }
            save(cals, to: url("calibrations.json"))
            var subs: [Subject] = load(url("subjects.json"), default: [])
            subs.removeAll { $0.code == code }
            save(subs, to: url("subjects.json"))
        }
    }

    // MARK: - Sessions

    func createSession(_ s: Session) {
        queue.sync {
            var all: [Session] = load(url("sessions.json"), default: [])
            all.removeAll { $0.id == s.id }
            all.append(s)
            save(all, to: url("sessions.json"))
        }
    }

    func endSession(_ id: String, endedAt: String) {
        queue.sync {
            var all: [Session] = load(url("sessions.json"), default: [])
            if let idx = all.firstIndex(where: { $0.id == id }) {
                all[idx].endedAt = endedAt
                save(all, to: url("sessions.json"))
            }
        }
    }

    func getSession(_ id: String) -> Session? {
        queue.sync {
            let all: [Session] = load(url("sessions.json"), default: [])
            return all.first { $0.id == id }
        }
    }

    /// Sessions, newest first (by startedAt).
    func listSessions() -> [Session] {
        queue.sync {
            let all: [Session] = load(url("sessions.json"), default: [])
            return all.sorted { $0.startedAt > $1.startedAt }
        }
    }

    func deleteSession(_ id: String) {
        queue.sync {
            var sessions: [Session] = load(url("sessions.json"), default: [])
            sessions.removeAll { $0.id == id }
            save(sessions, to: url("sessions.json"))
            var annos: [Annotation] = load(url("annotations.json"), default: [])
            annos.removeAll { $0.sessionId == id }
            save(annos, to: url("annotations.json"))
            try? fm.removeItem(at: chunksDir.appendingPathComponent("\(id).json"))
            try? fm.removeItem(at: metricsDir.appendingPathComponent("\(id).json"))
        }
    }

    // MARK: - Annotations

    @discardableResult
    func addAnnotation(_ a: Annotation) -> Int {
        queue.sync {
            var all: [Annotation] = load(url("annotations.json"), default: [])
            var item = a
            item.id = nextId(all.map { $0.id })
            all.append(item)
            save(all, to: url("annotations.json"))
            return item.id!
        }
    }

    func listAnnotations(_ sessionId: String) -> [Annotation] {
        queue.sync {
            let all: [Annotation] = load(url("annotations.json"), default: [])
            return all.filter { $0.sessionId == sessionId }.sorted { $0.tUs < $1.tUs }
        }
    }

    // MARK: - Calibrations

    @discardableResult
    func saveCalibration(_ c: Calibration) -> Int {
        queue.sync {
            var all: [Calibration] = load(url("calibrations.json"), default: [])
            var item = c
            item.id = nextId(all.map { $0.id })
            all.append(item)
            save(all, to: url("calibrations.json"))
            return item.id!
        }
    }

    func latestCalibration(_ subjectCode: String) -> Calibration? {
        listCalibrations(subjectCode).last
    }

    func listCalibrations(_ subjectCode: String) -> [Calibration] {
        queue.sync {
            let all: [Calibration] = load(url("calibrations.json"), default: [])
            return all.filter { $0.subjectCode == subjectCode }.sorted { $0.createdAt < $1.createdAt }
        }
    }

    // MARK: - Metrics

    func addMetrics(_ m: MetricsRecord) {
        queue.sync {
            let file = metricsDir.appendingPathComponent("\(m.sessionId).json")
            var all: [MetricsRecord] = load(file, default: [])
            var item = m
            item.id = nextId(all.map { $0.id })
            all.append(item)
            save(all, to: file)
        }
    }

    func listMetrics(_ sessionId: String) -> [MetricsRecord] {
        queue.sync {
            let file = metricsDir.appendingPathComponent("\(sessionId).json")
            let all: [MetricsRecord] = load(file, default: [])
            return all.sorted { $0.tUs < $1.tUs }
        }
    }

    // MARK: - Stream chunks

    func addChunk(_ chunk: StreamChunk) {
        queue.sync {
            let file = chunksDir.appendingPathComponent("\(chunk.sessionId).json")
            var all: [StreamChunk] = load(file, default: [])
            var item = chunk
            item.id = nextId(all.map { $0.id })
            all.append(item)
            save(all, to: file)
        }
    }

    private func chunks(_ sessionId: String) -> [StreamChunk] {
        let file = chunksDir.appendingPathComponent("\(sessionId).json")
        let all: [StreamChunk] = load(file, default: [])
        return all.sorted { $0.tStartUs < $1.tStartUs }
    }

    /// Concatenate stored chunks back into contiguous per-stream arrays.
    func readSessionStreams(_ sessionId: String) -> SessionStreams {
        queue.sync {
            let all = chunks(sessionId)
            var out = SessionStreams()
            for c in all where c.stream == .ecg {
                out.ecg.t += c.t
                out.ecg.ecg += c.ecg ?? []
            }
            for c in all where c.stream == .bioz {
                out.bioz.t += c.t
                out.bioz.z0 += c.z0 ?? []
                out.bioz.dz += c.dz ?? []
            }
            for c in all where c.stream == .ppg {
                out.ppg.t += c.t
                out.ppg.green += c.green ?? []
                out.ppg.red += c.red ?? []
                out.ppg.ir += c.ir ?? []
            }
            return out
        }
    }

    /// Wipe every stored table (Settings → data management).
    func deleteAllData() {
        queue.sync {
            for name in ["subjects.json", "sessions.json", "calibrations.json", "annotations.json"] {
                try? fm.removeItem(at: url(name))
            }
            try? fm.removeItem(at: chunksDir)
            try? fm.removeItem(at: metricsDir)
            try? fm.createDirectory(at: chunksDir, withIntermediateDirectories: true)
            try? fm.createDirectory(at: metricsDir, withIntermediateDirectories: true)
        }
    }
}

// MARK: - Recorder

/// Buffers live samples and flushes them to the store in ~2 s chunks. One instance per
/// active recording session. Robust to mid-stream disconnects (flush on stop).
/// Mirrors the web `Recorder` (web/src/model/repository.ts).
final class Recorder {
    let sessionId: String
    private(set) var counts: [StreamKind: Int] = [.ecg: 0, .bioz: 0, .ppg: 0]

    private final class BufferState {
        var tStartUs: Double? = nil
        var rate = 0
        var t: [Double] = []
        var a: [Double] = [] // ecg samples | z0 | green
        var b: [Double] = [] // dz | red
        var c: [Double] = [] // ir
    }

    private var buffers: [StreamKind: BufferState] = [
        .ecg: BufferState(), .bioz: BufferState(), .ppg: BufferState(),
    ]
    private let flushSamples: [StreamKind: Int] = [.ecg: 512, .bioz: 128, .ppg: 800]
    private let store: Store

    init(sessionId: String, store: Store = .shared) {
        self.sessionId = sessionId
        self.store = store
    }

    func ingestEcg(_ p: EcgPacket) {
        let buf = buffers[.ecg]!
        if buf.tStartUs == nil { buf.tStartUs = p.tUs }
        buf.rate = p.sampleRateHz
        let dt = DSPConstants.usPerS / Double(p.sampleRateHz)
        for i in 0..<p.samples.count {
            buf.t.append(p.tUs + (Double(i) * dt).rounded())
            buf.a.append(Double(p.samples[i]))
        }
        counts[.ecg, default: 0] += p.samples.count
        if buf.a.count >= flushSamples[.ecg]! { flush(.ecg) }
    }

    func ingestBioz(_ p: BiozPacket) {
        let buf = buffers[.bioz]!
        if buf.tStartUs == nil { buf.tStartUs = p.tUs }
        buf.rate = p.sampleRateHz
        let dt = DSPConstants.usPerS / Double(p.sampleRateHz)
        for i in 0..<p.dz.count {
            buf.t.append(p.tUs + (Double(i) * dt).rounded())
            buf.a.append(Double(p.z0Milliohm)) // forward-filled z0
            buf.b.append(Double(p.dz[i]))
        }
        counts[.bioz, default: 0] += p.dz.count
        if buf.a.count >= flushSamples[.bioz]! { flush(.bioz) }
    }

    func ingestPpg(_ p: PpgPacket) {
        let buf = buffers[.ppg]!
        if buf.tStartUs == nil { buf.tStartUs = p.tUs }
        buf.rate = p.sampleRateHz
        let dt = DSPConstants.usPerS / Double(p.sampleRateHz)
        for i in 0..<p.green.count {
            buf.t.append(p.tUs + (Double(i) * dt).rounded())
            buf.a.append(Double(p.green[i]))
            buf.b.append(Double(p.red[i]))
            buf.c.append(Double(p.ir[i]))
        }
        counts[.ppg, default: 0] += p.green.count
        if buf.a.count >= flushSamples[.ppg]! { flush(.ppg) }
    }

    /// Persist buffered samples for one stream as a StreamChunk.
    func flush(_ stream: StreamKind) {
        let buf = buffers[stream]!
        guard !buf.a.isEmpty, let tStart = buf.tStartUs else { return }
        var chunk = StreamChunk(id: nil, sessionId: sessionId, stream: stream,
                                tStartUs: tStart, sampleRateHz: buf.rate, t: buf.t)
        switch stream {
        case .ecg:
            chunk.ecg = buf.a.map { Int32($0) }
        case .bioz:
            chunk.z0 = buf.a.map { Int32($0) }
            chunk.dz = buf.b.map { Int32($0) }
        case .ppg:
            chunk.green = buf.a.map { UInt32($0) }
            chunk.red = buf.b.map { UInt32($0) }
            chunk.ir = buf.c.map { UInt32($0) }
        }
        buffers[stream] = BufferState()
        store.addChunk(chunk)
    }

    func flushAll() {
        for s in StreamKind.allCases { flush(s) }
    }
}

File: android/app/src/main/java/com/monivitals/model/Store.kt
//
//  Store.kt
//  MoniVitals (Android)
//
//  Dependency-light, local-first persistence (no cloud sync). Mirrors the responsibilities
//  of the web Dexie database and the iOS JSON-file Store, storing data as JSON files under
//  the app's private files directory:
//
//      <filesDir>/MoniVitals/
//        subjects.json          [Subject]
//        sessions.json          [Session]
//        calibrations.json      [Calibration]
//        annotations.json       [Annotation]
//        chunks/<sessionId>.json     [StreamChunk]   (high-rate sample blocks)
//        metrics/<sessionId>.json    [MetricsRecord]
//
//  Per-session sample data lives in its own file so listing/deleting sessions never
//  loads every waveform. All access is serialized on a lock.
//
//  Ported from ios/MoniVitals/Model/Store.swift.
//

package com.monivitals.model

import android.content.Context
import kotlinx.serialization.json.Json
import java.io.File

/** Concatenated, contiguous streams for a session (replay / export). */
class SessionStreams {
    class Ecg { val t = mutableListOf<Double>(); val ecg = mutableListOf<Int>() }
    class Bioz { val t = mutableListOf<Double>(); val z0 = mutableListOf<Int>(); val dz = mutableListOf<Int>() }
    class Ppg {
        val t = mutableListOf<Double>()
        val green = mutableListOf<Long>()
        val red = mutableListOf<Long>()
        val ir = mutableListOf<Long>()
    }

    val ecg = Ecg()
    val bioz = Bioz()
    val ppg = Ppg()
}

class Store private constructor(baseDir: File) {
    private val lock = Any()
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true; prettyPrint = false }

    private val root = File(baseDir, "MoniVitals")
    private val chunksDir = File(root, "chunks")
    private val metricsDir = File(root, "metrics")

    init {
        root.mkdirs()
        chunksDir.mkdirs()
        metricsDir.mkdirs()
    }

    companion object {
        @Volatile private var instance: Store? = null

        fun init(context: Context) {
            if (instance == null) {
                synchronized(this) {
                    if (instance == null) instance = Store(context.filesDir)
                }
            }
        }

        val shared: Store
            get() = instance ?: error("Store.init(context) must be called before Store.shared")
    }

    // MARK: - file helpers

    private fun file(name: String) = File(root, name)

    private val subjectsList = ListSerializer(Subject.serializer())
    private val sessionsList = ListSerializer(Session.serializer())
    private val calibsList = ListSerializer(Calibration.serializer())
    private val annosList = ListSerializer(Annotation.serializer())
    private val chunksList = ListSerializer(StreamChunk.serializer())
    private val metricsList = ListSerializer(MetricsRecord.serializer())

    private fun <T> load(f: File, serializer: kotlinx.serialization.KSerializer<List<T>>, def: List<T>): List<T> {
        if (!f.exists()) return def
        return try {
            json.decodeFromString(serializer, f.readText())
        } catch (e: Exception) {
            def
        }
    }

    private fun <T> save(value: List<T>, serializer: kotlinx.serialization.KSerializer<List<T>>, f: File) {
        try {
            val tmp = File(f.parentFile, f.name + ".tmp")
            tmp.writeText(json.encodeToString(serializer, value))
            tmp.renameTo(f)
        } catch (e: Exception) {
            // best-effort local store
        }
    }

    private fun nextId(items: List<Int?>): Int = (items.mapNotNull { it }.maxOrNull() ?: 0) + 1

    // MARK: - Subjects

    fun upsertSubject(s: Subject) = synchronized(lock) {
        val all = load(file("subjects.json"), subjectsList, emptyList()).toMutableList()
        all.removeAll { it.code == s.code }
        all.add(s)
        save(all, subjectsList, file("subjects.json"))
    }

    /** Subjects, newest first (by createdAt). */
    fun listSubjects(): List<Subject> = synchronized(lock) {
        load(file("subjects.json"), subjectsList, emptyList()).sortedByDescending { it.createdAt }
    }

    fun getSubject(code: String): Subject? = synchronized(lock) {
        load(file("subjects.json"), subjectsList, emptyList()).firstOrNull { it.code == code }
    }

    fun deleteSubject(code: String) {
        // Delete the subject's sessions + calibrations, then the subject.
        val sessions = listSessions().filter { it.subjectCode == code }
        for (s in sessions) deleteSession(s.id)
        synchronized(lock) {
            val cals = load(file("calibrations.json"), calibsList, emptyList()).toMutableList()
            cals.removeAll { it.subjectCode == code }
            save(cals, calibsList, file("calibrations.json"))
            val subs = load(file("subjects.json"), subjectsList, emptyList()).toMutableList()
            subs.removeAll { it.code == code }
            save(subs, subjectsList, file("subjects.json"))
        }
    }

    // MARK: - Sessions

    fun createSession(s: Session) = synchronized(lock) {
        val all = load(file("sessions.json"), sessionsList, emptyList()).toMutableList()
        all.removeAll { it.id == s.id }
        all.add(s)
        save(all, sessionsList, file("sessions.json"))
    }

    fun endSession(id: String, endedAt: String) = synchronized(lock) {
        val all = load(file("sessions.json"), sessionsList, emptyList()).toMutableList()
        val idx = all.indexOfFirst { it.id == id }
        if (idx >= 0) {
            all[idx] = all[idx].copy(endedAt = endedAt)
            save(all, sessionsList, file("sessions.json"))
        }
    }

    fun getSession(id: String): Session? = synchronized(lock) {
        load(file("sessions.json"), sessionsList, emptyList()).firstOrNull { it.id == id }
    }

    /** Sessions, newest first (by startedAt). */
    fun listSessions(): List<Session> = synchronized(lock) {
        load(file("sessions.json"), sessionsList, emptyList()).sortedByDescending { it.startedAt }
    }

    fun deleteSession(id: String) = synchronized(lock) {
        val sessions = load(file("sessions.json"), sessionsList, emptyList()).toMutableList()
        sessions.removeAll { it.id == id }
        save(sessions, sessionsList, file("sessions.json"))
        val annos = load(file("annotations.json"), annosList, emptyList()).toMutableList()
        annos.removeAll { it.sessionId == id }
        save(annos, annosList, file("annotations.json"))
        File(chunksDir, "$id.json").delete()
        File(metricsDir, "$id.json").delete()
    }

    // MARK: - Annotations

    fun addAnnotation(a: Annotation): Int = synchronized(lock) {
        val all = load(file("annotations.json"), annosList, emptyList()).toMutableList()
        val item = a.copy(id = nextId(all.map { it.id }))
        all.add(item)
        save(all, annosList, file("annotations.json"))
        item.id!!
    }

    fun listAnnotations(sessionId: String): List<Annotation> = synchronized(lock) {
        load(file("annotations.json"), annosList, emptyList())
            .filter { it.sessionId == sessionId }
            .sortedBy { it.tUs }
    }

    // MARK: - Calibrations

    fun saveCalibration(c: Calibration): Int = synchronized(lock) {
        val all = load(file("calibrations.json"), calibsList, emptyList()).toMutableList()
        val item = c.copy(id = nextId(all.map { it.id }))
        all.add(item)
        save(all, calibsList, file("calibrations.json"))
        item.id!!
    }

    fun latestCalibration(subjectCode: String): Calibration? = listCalibrations(subjectCode).lastOrNull()

    fun listCalibrations(subjectCode: String): List<Calibration> = synchronized(lock) {
        load(file("calibrations.json"), calibsList, emptyList())
            .filter { it.subjectCode == subjectCode }
            .sortedBy { it.createdAt }
    }

    // MARK: - Metrics

    fun addMetrics(m: MetricsRecord) = synchronized(lock) {
        val f = File(metricsDir, "${m.sessionId}.json")
        val all = load(f, metricsList, emptyList()).toMutableList()
        val item = m.copy(id = nextId(all.map { it.id }))
        all.add(item)
        save(all, metricsList, f)
    }

    fun listMetrics(sessionId: String): List<MetricsRecord> = synchronized(lock) {
        val f = File(metricsDir, "$sessionId.json")
        load(f, metricsList, emptyList()).sortedBy { it.tUs }
    }

    // MARK: - Stream chunks

    fun addChunk(chunk: StreamChunk) = synchronized(lock) {
        val f = File(chunksDir, "${chunk.sessionId}.json")
        val all = load(f, chunksList, emptyList()).toMutableList()
        val item = chunk.copy(id = nextId(all.map { it.id }))
        all.add(item)
        save(all, chunksList, f)
    }

    private fun chunks(sessionId: String): List<StreamChunk> {
        val f = File(chunksDir, "$sessionId.json")
        return load(f, chunksList, emptyList()).sortedBy { it.tStartUs }
    }

    /** Concatenate stored chunks back into contiguous per-stream arrays. */
    fun readSessionStreams(sessionId: String): SessionStreams = synchronized(lock) {
        val all = chunks(sessionId)
        val out = SessionStreams()
        for (c in all.filter { it.stream == com.monivitals.ble.StreamKind.ECG }) {
            out.ecg.t += c.t
            out.ecg.ecg += c.ecg ?: emptyList()
        }
        for (c in all.filter { it.stream == com.monivitals.ble.StreamKind.BIOZ }) {
            out.bioz.t += c.t
            out.bioz.z0 += c.z0 ?: emptyList()
            out.bioz.dz += c.dz ?: emptyList()
        }
        for (c in all.filter { it.stream == com.monivitals.ble.StreamKind.PPG }) {
            out.ppg.t += c.t
            out.ppg.green += c.green ?: emptyList()
            out.ppg.red += c.red ?: emptyList()
            out.ppg.ir += c.ir ?: emptyList()
        }
        out
    }

    /** Wipe every stored table (Settings → data management). */
    fun deleteAllData() = synchronized(lock) {
        for (name in listOf("subjects.json", "sessions.json", "calibrations.json", "annotations.json")) {
            file(name).delete()
        }
        chunksDir.deleteRecursively()
        metricsDir.deleteRecursively()
        chunksDir.mkdirs()
        metricsDir.mkdirs()
    }

    // Exposed for the CSV exporter (single-source access to the per-session dirs).
    internal fun chunksDir() = chunksDir
    internal fun metricsDir() = metricsDir
}

private fun <T> ListSerializer(elem: kotlinx.serialization.KSerializer<T>) =
    kotlinx.serialization.builtins.ListSerializer(elem)
